### ISSUE TITLE
Benchmark replacing maps with unordered_maps

### ISSUE THREAD

**Issue by jpfeuffer** [OpenMS/OpenMS#6295]
Especially on ID data algorithms.
Maps with strings as keys, many entries, and many lookups.

For prioritizing, some internal knowledge might be needed but it can also be done just tool by tool.

**Comment by gaurav-iitb** [comment]
is it still open

**Comment by jpfeuffer** [comment]
Yes but please note that this issue is about benchmarking (i.e., running and timing multiple tools and algorithms after changing the code).
It is not hard to do but will take some time.

**Comment by cbielow** [comment]
... and input data ( you can ask us if you are not sure what to use)

**Comment by Telomelonia** [comment]
> Especially on ID data algorithms. Maps with strings as keys, many entries, and many lookups.
> 
> For prioritizing, some internal knowledge might be needed but it can also be done just tool by tool.

@jpfeuffer @cbielow 

I am thinking to grep for "map<" and changing it to "unordered_map<". Also, I am planning to benchmark using chrono or some 3rd party software.
But I have queries:
1. What is the meaning of ID algorithms.
2. Only maps with string as key are subjected to these changes?
3. Is it possible to make a single method/.cpp file to show the benchmark of all the changes.?
4. From where I can get the data? (ig from tutorials maybe)

**Comment by ProgrammingPirates** [comment]
Hi @cbielow 

I hope you’re doing well. I’ve carefully reviewed this issue and would like to request assignment so I can start contributing. I’m confident I can implement a proper solution while following the project’s coding standards and contribution guidelines.

I’m an aspiring contributor preparing for Google Summer of Code 2026, and I’m eager to learn and make meaningful contributions to this organization. Could you please assign this issue to me?

Thank you for your time and consideration. 🙏

Best regards,
Dharmendra Solanki

**Comment by khushiiagrawal** [comment]
hii @jpfeuffer, greetings.
I would like to work upon this issue. Is this issue still open to work on ?

**Comment by jpfeuffer** [comment]
Yes still open. Depending on what kind of parts you are tackling first, you might need to request some test data to benchmark on.
Maybe you can also let AI generate some random data directly into your data structures (which might be more independently testable but maybe also not completely representative).

**Comment by NarendraRaoJami** [comment]
Hi @jpfeuffer , I’m new to OpenMS and would like to work on this issue.
I plan to benchmark selected std::map usages against std::unordered_map and report results.

**Comment by im-alok74** [comment]
Hi, I’ve built OpenMS locally and plan to start by benchmarking one ID tool (e.g., IDScoreSwitcher) by comparing std::map<string> vs std::unordered_map<string>. I’ll report timings before making broader changes. Please let me know if this approach works or if you recommend a specific dataset

**Comment by yuvraj-coder1** [comment]
hi @jpfeuffer,
I built OpenMS locally and benchmarked one of the ID-related map usages in `PeptideIndexing.cpp`, specifically the acc_to_prot map (map<String, Size>). 

i used test data to compare map vs unordered_map performance.
### Used two different test data :-
1) Used data from `SuffixArrayPeptideFinder_test.fasta`
2) Generated accessions of the form:
```
sp|<index>|PROT_<index>
```
For example
```
sp|1|PROT_1
sp|2|PROT_2
sp|3|PROT_3
...  
```

Tested dataset sizes:

- 1k proteins  
- 10k proteins  
- 100k proteins  

Each benchmark runs **10,000 iterations** with mixed inserts (~50%) and lookups (~33%) to expose scaling behavior.
(10,000 iterations is a stress test, for the 100k dataset this implies around 8.3×10^8 map operations)

### Result

For `SuffixArrayPeptideFinder_test.fasta` 
- `std::map` → ~4.1 s  
- `std::unordered_map` → ~2.1 s  
- **≈ 1.9× speedup**

For generated dataset

| Size  | std::map | std::unordered_map | Speedup |
|-------|----------|--------------------|---------|
| 1k    | ~4.7 s   | ~1.9 s             | ~2.4×   |
| 10k   | ~61.8 s  | ~25.1 s            | ~2.5×   |
| 100k  | ~1076 s  | ~576 s             | ~1.9×   |


should i benchmark other ID-related maps similarly or use a different dataset or method?

**Comment by jpfeuffer** [comment]
@yuvraj-coder1 that looks nice. You can also make a PR after testing certain occurrences and then continue with the next one.
Sometimes you have to be careful because the tests might depend on the ordering in the map.
In that case one must check if the algorithm can be changed to not depend on ordering or sorting must be performed in testing mode. Or if the speedup is not significant, map should be kept.

**Comment by krrishverma1805-web** [comment]
Hi! I noticed that PR #8763 successfully replaced `std::map` with `std::unordered_map` in `PeptideIndexing.cpp` and was merged recently. I'd like to continue this effort by identifying and benchmarking additional files where `std::map` with string keys can be safely replaced with `std::unordered_map`. 

I'll focus on ID-related data structures as mentioned in the original issue description, make sure ordering isn't relied upon, and include benchmark results. 

Could I pick up another file for this? I'll start with analysis and share my findings before making changes.

**Comment by yuvraj-coder1** [comment]
hi @krrishverma1805-web thanks for the initiative, I’m still actively working on extending this across additional files and benchmarking the changes to ensure there’s an actual performance improvement and that ordering assumptions aren’t affected. I’ll share updates once I have consolidated results.

**Comment by SanketMeghale** [comment]
Hi @jpfeuffer, I would like to work on this issue with a focused first PR.\n\nPlanned scope for PR-1:\n1. Pick one additional ID-related std::map<String, ...> occurrence (not already changed in #8763)\n2. Benchmark before/after (map vs unordered_map) with reproducible input and report timings\n3. Verify ordering assumptions and keep map where ordering is required\n4. Submit a small PR with tests passing and benchmark summary\n\nIf this scope looks good, I’ll start with candidate analysis and share the exact file/function before opening the PR.

**Comment by SanketMeghale** [comment]
Quick update: I started with IDMapper as first focused candidate.\n\nConcrete location:\n- src/openms/source/ANALYSIS/ID/IDMapper.cpp\n- TMT/iTRAQ mapping cache ile2nativeid2pepid inside IDMapper::annotate(ConsensusMap&, ...)\n\nThis cache is used as key-based lookup (ind) by filename + scan id and does not rely on key iteration order, so it is a good candidate to benchmark std::map vs std::unordered_map safely.\n\nI’ll report before/after timings next for this location.

**Comment by nej1gotnochill** [comment]
Hi @jpfeuffer,

I'd like to contribute further to this issue.

Before commenting, I went through the original discussion as well as the merged PRs **#8763** and **#8806** to understand the expected workflow and selection criteria.

My plan is to:

1. Identify another suitable occurrence of `std::map<String, ...>` (or `std::map<std::string, ...>`) that has not yet been optimized.
2. Verify that the container is primarily used for lookups and that iteration order is not relied upon.
3. Benchmark the existing implementation against `std::unordered_map` using representative data.
4. Submit a focused PR only if the change provides a measurable performance improvement without affecting correctness or introducing ordering-related regressions.

If no meaningful benefit is observed, I'll document the findings rather than making the replacement.

Does this approach align with your expectations for this issue? Also, are there any particular components or files you would recommend investigating next?

Thank you and have a nice day!

**Comment by nej1gotnochill** [comment]
Hi @timosachsenberg, thanks for your reply on discord ! I was referring to PR #8763 and PR #8806, which I had mentioned in my earlier comment. I reviewed both to understand the previous std::map → std::unordered_map replacements and planned to follow the same approach by identifying another suitable occurrence, benchmarking it, and opening a similarly scoped PR if the change provides a measurable benefit.
